File: swift/0078-subsets.swift
class Solution {
    func subsets(_ nums: [Int]) -> [[Int]] {
        var res : [[Int]] = []
        var subset : [Int] = []
        
        func dfs(_ i: Int) {
            if i >= nums.count {
                res.append(subset)
                return
            }
            // decision to include nums[i]
            subset.append(nums[i])
            dfs(i + 1)
            // decision NOT to include nums[i]
            subset.removeLast()
            dfs(i + 1)
        }
        
        dfs(0)
        return res
    }
}

File: csharp/0078-subsets.cs
public class Solution {
    private List<int> subset = new List<int>();
    private List<IList<int>> result = new List<IList<int>>();
    private void backtrack(int i, int[] nums) {
        if(i >= nums.Length) {
            result.Add(new List<int>(subset));
            return;
        }
        subset.Add(nums[i]);
        backtrack(i + 1, nums);
        subset.Remove(nums[i]);
        backtrack(i + 1, nums);
            
    }
    public IList<IList<int>> Subsets(int[] nums) {
        backtrack(0, nums);
        return result;  
    }
}